src/siwe.rs

```diff
@@ -0,0 +1,357 @@
+use crate::{
+    auth::{Action, AuthorizationPolicy, AuthorizationToken},
+    orbit::{hash_same, OrbitMetadata},
+    zcap::KeplerInvocation,
+};
+use anyhow::Result;
+use didkit::DID_METHODS;
+use ipfs_embed::Cid;
+use rocket::{
+    http::Status,
+    request::{FromRequest, Outcome, Request},
+};
+
+use ethers_core::{types::H160, utils::to_checksum};
+use hex::FromHex;
+use serde::{Deserialize, Serialize};
+use serde_with::{serde_as, DisplayFromStr};
+use siwe::eip4361::Message;
+use std::{ops::Deref, str::FromStr};
+
+pub struct SIWESignature([u8; 65]);
+
+impl core::fmt::Display for SIWESignature {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        write!(f, "0x{}", hex::encode(&self.0))
+    }
+}
+
+impl FromStr for SIWESignature {
+    type Err = anyhow::Error;
+    fn from_str<'a>(s: &'a str) -> Result<Self, Self::Err> {
+        match s.split_once("0x") {
+            Some(("", h)) => Ok(Self(<[u8; 65]>::from_hex(h)?)),
+            _ => Err(anyhow!("Invalid hex string, no leading 0x")),
+        }
+    }
+}
+
+#[serde_as]
+#[derive(Serialize, Deserialize)]
+pub struct SIWEMessage(
+    #[serde_as(as = "DisplayFromStr")] Message,
+    #[serde_as(as = "DisplayFromStr")] SIWESignature,
+);
+
+impl Deref for SIWEMessage {
+    type Target = Message;
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+pub struct SIWEZcapTokens {
+    pub invocation: KeplerInvocation,
+    pub delegation: SIWEMessage,
+}
+
+pub struct SIWETokens {
+    pub invocation: SIWEMessage,
+    pub delegation: Option<SIWEMessage>,
+    // kinda weird
+    pub orbit: Cid,
+    pub action: Action,
```

Comment:
The SIWEMessage can delegate multiple actions, but only the first in the list is stored in this field (line 132). What's the deal with that?

Reply:
the `action` in this struct is the one being invoked, not delegated, but yeah it's not ideal, it expects that an invokation will contain just one action (which is not guarenteed anywhere else, the rest will just be ignored).

Reply:
Ok that makes sense, but it should be documented that the other resources are ignored in an invocation, and it might also be clearer if this field is renamed to `invoked_action`.

Reply:
for context, it's undecided right now if we should or shouldn't allow multiple actions/resources in an invocation. allowing such would make designing a HTTP api tricky (what is the appropriate route for an inv containing both GET and PUT actions?) and/or opens questions about implicitly allowing re-use of an invocation (could you use such a token on both GET and PUT routes? generally this is considered a bad idea). IMO we should reject them